### PR TITLE
Bugfix: remove deprecated GoogleSignIn and refactor sign in

### LINKED ISSUE #177: Auth: fix sign in bugs

### PR DESCRIPTION
**Pull request by santiagoarayaramos**
# Description

Completely removed the use of the deprecated GoogleSignIn, using Credential Manager instead. This also involved a lot of refactoring to increase readability of the code, and move most code to the renamed Authenticator file (instead of defining it in Sign In). This now handles sign in with credential manager (using google's federated identity), and authentication with Firebase Auth.

The Authenticator is also moved out of the sign in class to enable correct logout (including the resetting of the credential manager state) in profile screens.

Finally, replaced the creation of unmanaged unconnected coroutines with safer coroutines connected to the composable functions they are called in.

Fixes #177 

# How Has This Been Tested?

Existing tests were refactored to fit new function signatures, and new tests were added.

### Unit Tests
Tests for Authenticator class were added, as much as possible but with many static final functions and classes that use obscure and not easily creatable Bundles and classes, testing is hard.

  
# Dependencies

No new dependencies added

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged in downstream modules, no conflicts with main

### CODE REVIEW

File: app/src/main/java/com/github/onlynotesswent/ui/user/PublicProfile.kt [hunk 1]
```diff
@@ -126,15 +130,16 @@ fun PublicProfileScreen(
     navigationActions: NavigationActions,
     userViewModel: UserViewModel,
     fileViewModel: FileViewModel,
-    notificationViewModel: NotificationViewModel
+    notificationViewModel: NotificationViewModel,
+    authenticator: Authenticator,
```
Comment (aaisel) [inline app/src/main/java/com/github/onlynotesswent/ui/user/PublicProfile.kt:134]: Update comments 😸

File: app/src/main/java/com/github/onlynotesswent/ui/user/PublicProfile.kt [hunk 2]
```diff
@@ -157,6 +162,7 @@ private fun ProfileScaffold(
     navigationActions: NavigationActions,
     userViewModel: UserViewModel,
     notificationViewModel: NotificationViewModel,
+    authenticator: Authenticator,
```
Comment (aaisel) [inline app/src/main/java/com/github/onlynotesswent/ui/user/PublicProfile.kt:165]: Update comments 😆

File: app/src/main/java/com/github/onlynotesswent/ui/user/PublicProfile.kt [hunk 3]
```diff
@@ -200,6 +207,7 @@ private fun ProfileScaffold(
  * @param title The title to be displayed on the app bar.
  * @param navigationActions The navigation actions.
  * @param userViewModel The ViewModel for the user.
+ * @param authenticator The Authenticator used for the app, to enable sign out of the user.
```
Comment (aaisel) [inline app/src/main/java/com/github/onlynotesswent/ui/user/PublicProfile.kt:210]: Awesome job at updating the doc 🎊

File: app/src/test/java/com/github/onlynotesswent/model/authentication/AuthenticatorTest.kt [hunk 4]
```diff
@@ -0,0 +1,428 @@
...
+private fun verifyErrorLog(msg: String) {
+  // Get all the logs
+  val logs = ShadowLog.getLogs()
+
+  // Check for the debug log that should be generated
+  val errorLog = logs.find { it.type == Log.ERROR && it.tag == Authenticator.TAG && it.msg == msg }
+  assert(errorLog != null) { "Expected error log was not found!" }
+}
+
+// package com.github.onlynotesswent.model.authentication
```
Comment (aaisel) [inline app/src/test/java/com/github/onlynotesswent/model/authentication/AuthenticatorTest.kt:235]: As mentioned above, testing sign in is difficult. Coverage is good enough, tests not used and commented out could be removed

**Review by aaisel — CHANGES_REQUESTED** [review]
Overall good job at tackling the previous issues that we had. Your code is well structured and easy to understand, comments are clear and useful. We can merge this even if the coverage is not enough - testing sign in is difficult, and there is no point in wasting a lot of time on it. 
I only requested a few minor changes about updating the comments and removing some Toasts.

File: app/src/main/java/com/github/onlynotesswent/ui/user/PublicProfile.kt [hunk 5]
```diff
@@ -86,7 +88,8 @@ fun UserProfileScreen(
     navigationActions: NavigationActions,
     userViewModel: UserViewModel,
     fileViewModel: FileViewModel,
-    notificationViewModel: NotificationViewModel
+    notificationViewModel: NotificationViewModel,
+    authenticator: Authenticator,
```
[the remark this answers is not in the document]
Comment (santiagoarayaramos) [reply app/src/main/java/com/github/onlynotesswent/ui/user/PublicProfile.kt:92]: Nice catch!!

**Comment by aaisel** [comment]
Great job at addressing requested changes!

**Pull request merged**